README.md

```diff
@@ -1,2 +1,89 @@
 # app.js
-GitHub Apps toolset for Node.js
+
+> Official GitHub Platform for GitHub Apps in Node.js
+
+## GitHub Apps
+
+`@octokit/app.js` allows to authenticate as GitHub Apps, installations, listen
+to webhooks, and respond to them using the GitHub API.
+
+### Authenticating as an App
+
+In order to authenticate as a GitHub App, you need to generate a Private Key
+and use it to sign a JSON Web Token (jwt) and encode it. 
+
+```js
+const App = require('@octokit/app.js')
+
+const app = new App({id, privateKey})
+const appAuth = await app.appAuth()
+
+// Example of using authetincated app to GET an individual installation
+// https://developer.github.com/v3/apps/#find-repository-installation
+const installation = await appAuth.request({
+  method: 'GET',
+  url: '/repos/:owner/:repo/installation',
+  headers: { accept: 'application/vnd.github.machine-man-preview+json' },
+  owner: 'hiimbex',
+  repo: 'testing-things'
+})
+
+// contains the installation id necessary to authenticate as an installation
+const installationId = installation.data.id
+```
+
+### Authenticating as an Installation
+
+Once you have authenticated as a GitHub App, you can use that
+in order to request an installation access token. Calling `requestToken()`
+automatically performs the app authentication for you. This token is scoped for
+your specific app and expires after an hour. 
+
+```js
+const app = require('@octokit/app.js')
+
+const app = new App({id, privateKey})
+const installation = await app.requestToken({installationId: 123})
+
+// Example of opening an issue as an installation
+// https://developer.github.com/v3/issues/#create-an-issue
+const result = await installation.issues.create({
+  owner: 'hiimbex',
+  repo: 'tetsing-things',
+  title: 'My installation’s first issue!'
+})
+```
```

Comment:
This is really interesting! I didn’t think of getting a client API from `app.appAuth()` or `app.requestToken()`, I thought they would only return the tokens. 

I’m not sure what the better approach is here. 

The advantage of only returning the token is that the `@octokit/app` package will be much simpler and less opinionated. The user can use their own request library, I was thinking of something like this:

```js
const got = require('got') // could be any request library really
const app = require('@octokit/app.js')
const installationAccessToken = await app.requestToken({installationId: 123})

// https://developer.github.com/v3/issues/#create-an-issue
const {data: {total_count}} = await got.post('https://api.github.com/repos/hiimbex/tetsing-things/issues', {
  body: {title: 'My installation’s first issue'},
  headers: {
    authorization: `token ${installationAccessToken}`,
    accept: 'application/vnd.github.machine-man-preview+json'
  }
})
```

I’m curious to hear what you think! 

Maybe as contrast, I really like your approach for the  "all-batteries-included" `octokit` npm package. The `octokit` package would have an `.app()` method which would use `@octokit/app` internally, but it would also have all the REST and GraphQL APIs etc. I could imagine an API like this

```js
const Octokit = require('octokit')
const client = new Octokit()

const app = client.app({id, privateKey})
const installation = app.findRepoInstallation({owner, repo})
```

We wouldn’t even need to call a method like `app.appAuth()` to get the JWT token, it could all happen internally when needed, because we have the app id and privateKey as internal state. When the user calls `.findRepoInstallation()` than the `app` instance could check if it has a cached JWT token that didn’t time out yet. If it does, it uses that token to send the actual request. Otherwise it will send the request to receive the JWT token first, cache it for future requests, and only then send the actual request to create the issue.

Reply:
>The advantage of only returning the token is that the @octokit/app package will be much simpler and less opinionated.

Totally agree with this sentiment.

I really like the idea of an "all-parts-included" octokit to make this easy. As you saw, I kept doing it even accidentally when making this write-up! I feel like the equivalent of `context.github` would be really nice (ie having both rest and graphql built in)

I think a lot of folks would prefer something that does all the work for them (see: the success of probot) as opposed to having to combine all the parts together. And I'm not totally convinced yet that there's a _need_ for a simple `@octokit/app.js` package that doesn't include this sort of built together stuff. When you consider that Octokit.js is officially supported by GitHub, I'm not sure we need to be that flexible in our "opinionation" of this sort of library.

Reply:
My thinking is to recommend everyone to use the `octokit` module if they want to integrate with GitHub, no matter if they want to build a GitHub App, an OAuth App, send requests to the REST API, GraphQL API or just to receive webhooks. The `@octokit/app` module could be used directly if people need lower-level control, just like with `@octokit/rest`, `@otokit/webhooks`, etc.

Reasons to use the separate modules could be constraints like hard drive or memory space.

The current example code could look something like this when using the `octokit` module instead of `@octokit/app`

```js
const Octokit = require('octokit')
const client = new Octokit({app: {id, privateKey}})
const installation = client.app.asInstallation({installationId})

// Example of opening an issue as an installation
// https://developer.github.com/v3/issues/#create-an-issue
const result = await installation.rest.issues.create({
  owner: 'hiimbex',
  repo: 'tetsing-things',
  title: 'My installation’s first issue!'
})
```

Internally, the `.asInstallation()` method would setup `@ocotkit/rest` and `@octokit/graphql` so that it always has a valid installation token. The `installation` instance would have the `.rest` method with the API of `@ocotkit/rest` and a `.graphql` method that will use a yet to be created `@octokit/graphql` method.

If it helps I’d be happy to chat, or we can also discuss https://github.com/octokit/octokit.js/pull/2 at the same time to see how the two modules would ideally work together?

Reply:
👍 I think this perfectly describes the best case scenario!

Note: would `const Octokit = require('octokit')` actually be `const Octokit = require('octokit.js')`?

But yea let's set up a time to chat!